src/grep.rs

```diff
@@ -26,19 +27,21 @@ fn build_pattern(
 
 /// Run bgrep, outputting `path` to the given `StdoutLock` if there is a match.
 /// Returns whether there was a match.
-fn grep_filename(
-  stdout: &mut io::StdoutLock,
+fn grep_filename<O: Write, P: Display, B: AsRef<[u8]>>(
+  stdout: &mut O,
   options: &args::Options,
   pattern: &Regex,
-  path: path::Display,
-  buffer: &[u8]
+  path: P,
+  buffer: B
 ) -> io::Result<bool> {
+  let buffer = buffer.as_ref();
+
   // When inverse matching, matches must be checked until a "hole" is found.
   // Otherwise, the more performant `Regex::is_match` can be used.
   if options.inverse {
     // if the pattern matches multiple times, comprising the entire buffer, then no
     // inverse match is present.
-    let mut matches = pattern.find_iter(buffer);
+    let matches = &mut pattern.find_iter(buffer);
```

Comment:
I believe a reference has no benefit here. If I'm right, could you revert this please?

Reply:
Ah, they are equivalent in this case (and cases below). IMHO, the advantage of using `let matches = &mut ...` over `let mut matches = ...` is restrict the mutability: we don't need to use a mutable binding if we don't want to rebind it.

I'm happy to change it if you don't want, actually it's just "good//bad practices" and, as you've said, has no real benefit.

Thanks a lot for your kindness.

Reply:
But how does `&mut` prevent rebinding? From what I know, you can do this:
```rust
fn main() {
    let mut y = vec![5];
    y = vec![10];
    println!("{:?}", y);
    
    let x = &mut vec![5];
    *x = vec![10];
    println!("{:?}", x);
}
```

Reply:
IMHO, what we are doing in `*x = vec![10]` is to implicitly change `*x` via _dereference_, then it depends on the semantics of the dereference operator (some cases, it could be something which may be not familiar with us).

Such a modification is not "protected" by the type system (e.g. we can define a new dereference semantics via Deref), while the move `y = vec![10]` has a well-defined semantics.

So doing `*x = vec![10]` is implicit (IMHO), then there is less chance that we accidentally use it to change `x`, than the "well-defined" `y = vec![10]`.

This is just my gut feeling, so don't take it seriously. I've changed the PR anyway.

Reply:
Oh, I see now. I didn't think of the explicitness of `*x` vs. `x`, and it makes sense. The change is surely valid, thanks for showing your point of view. As you have already changed back the code, I believe we can have this in another PR if you'd like.